monitors/ContractMonitor.js

```diff
@@ -0,0 +1,173 @@
+const { Logger } = require("../financial-templates-lib/logger/Logger");
+
+const { createFormatFunction, createEtherscanLinkMarkdown } = require("../common/FormattingUtils");
+const networkUtils = require("../common/PublicNetworks");
+
+class ContractMonitor {
+  constructor(expiringMultiPartyEventClient, account, umaLiquidatorAddress, umaDisputerAddress) {
+    // Bot and ecosystem accounts.
+    this.account = account;
+    this.umaLiquidatorAddress = umaLiquidatorAddress;
+    this.umaDisputerAddress = umaDisputerAddress;
+
+    // Previous contract state used to check for new entries between calls
+    this.lastLiquidationBlockNumber = 0;
+    this.lastDisputeBlockNumber = 0;
+    this.lastDisputeSettlementBlockNumber = 0;
+
+    // EMP event client to read latest contract events
+    this.empEventClient = expiringMultiPartyEventClient;
+    this.web3 = this.empEventClient.web3;
+
+    // Contract constants
+    //TODO: replace this with an actual query to the collateral currency symbol
+    this.collateralCurrencySymbol = "DAI";
+    this.syntheticCurrencySymbol = "UMATEST";
+
+    // TODO: get the decimals of the collateral currency and use this to scale the output appropriately for non 1e18 colat
+    this.formatDecimalString = createFormatFunction(this.web3, 2);
+  }
+
+  // Calculate the collateralization Ratio from the collateral, token amount and token price
+  // This is cr = [collateral / (tokensOutstanding * price)] * 100
+  calculatePositionCRPercent = (collateral, tokensOutstanding, priceFunction) => {
+    return this.web3.utils
+      .toBN(collateral)
+      .mul(this.web3.utils.toBN(this.web3.utils.toWei("1")))
+      .mul(this.web3.utils.toBN(this.web3.utils.toWei("1")))
+      .div(this.web3.utils.toBN(tokensOutstanding).mul(this.web3.utils.toBN(priceFunction.toString())))
+      .muln(100);
+  };
+
+  // Queries disputable liquidations and disputes any that were incorrectly liquidated.
+  checkForNewLiquidations = async priceFunction => {
+    Logger.debug({
+      at: "ContractMonitor",
+      message: "Checking for new liquidation events",
+      price: priceFunction.toString(),
+      lastLiquidationBlockNumber: this.lastLiquidationBlockNumber
+    });
+
+    // Get the latest liquidation information.
+    let latestLiquidationEvents = this.empEventClient.getAllLiquidationEvents();
+
+    // Get liquidation events that are newer than the last block number we've seen
+    let newLiquidationEvents = latestLiquidationEvents.filter(event => event.blockNumber > this.lastDisputeBlockNumber);
+
+    for (let event of newLiquidationEvents) {
+      // Sample message:
+      // Liquidation alert: [ethereum address if third party, or “UMA” if it’s our bot]
+      // initiated liquidation for for [x][collateral currency]of sponsor collateral
+      // backing[n] tokens - sponsor collateralization was[y] %.  [etherscan link to txn]
+      const mrkdwn =
```

Comment:
joining strings like this is not ideal. Originally I had used `...${}...` notation but it become even more difficult to read. Any sugestions around a cleaner way of building up this string?

Reply:
Block formatting *could* help? Although that may only really be useful for multiline strings. With such a long message, though, it could make sense to separate it into different lines or sections.

See [here](https://github.com/UMAprotocol/protocol/blob/5711c55d75db38651c65b9b78ae134c66ea72502/core/scripts/Voting.js#L639-L643) for an example.

Another idea is constructing different cohesive sections separately as we do in the example above and then concatenating them at the end.

Reply:
this could help but I feel that all of these options will just move the ugliness to a later point. Building different sections then joining them is also a good option. Could have separate sentences that come together to form one message. For now I think I will leave it like this, get the tests in and then refactor at a later point as this is just a syntactical sugar.

Reply:
SGTM.